app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/LocalFileTransferActivity.java

```diff
@@ -0,0 +1,437 @@
+                    PERMISSION_REQUEST_CODE_COARSE_LOCATION);
+                return Unit.INSTANCE;
+              }
+            });
+      } else {
+        ActivityCompat.requestPermissions(this, new String[] { Manifest.permission.ACCESS_COARSE_LOCATION },
+            PERMISSION_REQUEST_CODE_COARSE_LOCATION);
+      }
+      return false;
+
+    } else {
+      return true; // Control reaches here: Either permission granted at install time, or at the time of request
+    }
+  }
+
+  private boolean checkExternalStorageWritePermission() { // To access and store the zims
+    if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
+        == PackageManager.PERMISSION_DENIED) {
+
+      if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
+        alertDialogShower.show(KiwixDialog.StoragePermissionRationale.INSTANCE,
+            new Function0<Unit>() {
+              @Override public Unit invoke() {
+                ActivityCompat.requestPermissions(LocalFileTransferActivity.this, new String[] { Manifest.permission.WRITE_EXTERNAL_STORAGE },
+                    PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS);
+                return Unit.INSTANCE;
+              }
+            });
+      } else {
+        ActivityCompat.requestPermissions(this, new String[] { Manifest.permission.WRITE_EXTERNAL_STORAGE },
+            PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS);
+      }
+      return false;
+
+    } else {
+      return true; // Control reaches here: Either permission granted at install time, or at the time of request
+    }
+  }
+
+  @Override
+  public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+      @NonNull int[] grantResults) {
+
+    if (grantResults[0] == PackageManager.PERMISSION_DENIED) {
+      switch (requestCode) {
+
+        case PERMISSION_REQUEST_CODE_COARSE_LOCATION: {
+          Log.e(TAG, "Location permission not granted");
+
+          showToast(this, R.string.permission_refused_location, Toast.LENGTH_LONG);
+          finish();
+          break;
+        }
+
+        case PERMISSION_REQUEST_CODE_STORAGE_WRITE_ACCESS: {
+          Log.e(TAG, "Storage write permission not granted");
+
+          showToast(this, R.string.permission_refused_storage, Toast.LENGTH_LONG);
+          finish();
+          break;
+        }
+
+        default: {
+          super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+          break;
+        }
+      }
+    }
+  }
+
+  private boolean isLocationServicesEnabled() {
+    LocationManager locationManager =
+        (LocationManager) getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
+    boolean gps_enabled = false;
+    boolean network_enabled = false;
+
+    try {
+      gps_enabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
```

Comment:
you don't need the variables, try returning the `isProviderEnabled` calls and at the end return false

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/cb944bb88f3809b3baee4613e21e92972122a775#diff-e1d59ae6e15dcb502e125fb6e243d15dR339)

Reply:
```
return (locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER) || locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER));
```
I am afraid this will not work the same as your original code, if the first `isProviderEnabled` throws an exception then it will hit the catch block and never run the 2nd `isProviderEnabled`

Reply:
I think keeping each in a separate try-catch like [so](https://github.com/kiwix/kiwix-android/pull/1276/commits/c1288e6766825782f1fc9c4a49a3a81b5a778c56) will work?

Reply:
ah yes, sorry my mistake, you are correct

Reply:
So a nicer way is to extract a method `isProviderEnabled()` and in this function return `isProviderEnabled(GPS_PROVIDER) || isProviderEnabled(NETWORK_PROVIDER)`. gets rid of these variables, stops you duplicating your try/catch.
You should also probably inject `LocationManager`

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/20e1eebf378a70f5aa24e96f48c44161a2c57d4c)
Could you have a look at the injection code for LocationManager ([here](https://github.com/kiwix/kiwix-android/pull/1276/commits/20e1eebf378a70f5aa24e96f48c44161a2c57d4c#diff-563b1a2baeed937649eb846e3078dfcfR98)) once? I hope I didn't miss anything

Reply:
Injection code is A OK